lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,300 @@
+
+      // Shape annotation on ShapedType type -> push back new type
+      if (isDynamicShapedType && shapeAttr) {
+        auto shape =
+            getShapeFromArrayAttr(shapeAttr, shapedType, funcOp.getLoc());
+        if (failed(shape)) return shape;
+        // save the newly inferred argument type
+        inferredInputTypes.emplace_back(
+            shapedType.clone(shape.value(), shapedType.getElementType()));
+        changed = true;
+
+      } else {
+        // Dynamic ShapedType but no {shape.shape} annotation
+        if (isDynamicShapedType && !shapeAttr) {
+          emitWarning(funcOp.getLoc(),
+                      "Found argument with dynamic shaped type but no shape "
+                      "annotation.");
+        }
+
+        // Shape annotation on non-ShapedType type
+        if (shapeAttr && !shapedType) {
+          emitWarning(
+              funcOp.getLoc(),
+              "Ignoring shape annotation on argument with non-ShapedType type");
+        }
+
+        inferredInputTypes.emplace_back(type);
+      }
+    }
+
+    /////////////////////////////////////////////////////////////////////
+    //// Step 2: Walk the body, inferring and updating new types     ////
+    /////////////////////////////////////////////////////////////////////
+
+    // Only do this if there's been an annotated type and there's a body
+    if (changed && !funcOp.isDeclaration()) {
+      rewriter.startOpModification(funcOp);
+
+      //// Update Block Argument types based on Annotations
+      for (auto [idx, type] : llvm::enumerate(inferredInputTypes)) {
+        funcOp.getArgument(idx).setType(type);
+      }
+
+      // WalkOrder::PreOrder is essential so that we can handle mapping
+      // operands to block arguments for region-bearing ops before
+      // the region itself is walked.
+      auto walkResult =
+          funcOp.getBody().walk<WalkOrder::PreOrder>([&](Operation *op) {
+            llvm::dbgs() << "walking op: " << op->getName() << "\n";
+
+            if (auto regionBranchOp = dyn_cast<RegionBranchOpInterface>(op)) {
+              llvm::dbgs() << "found region branch op: " << op->getName()
+                           << "\n";
+
+              SmallVector<Attribute> operandAttributes(
+                  regionBranchOp->getNumOperands());
+              SmallVector<RegionSuccessor> successors;
+              regionBranchOp.getEntrySuccessorRegions(operandAttributes,
+                                                      successors);
+              for (auto &successor : successors) {
+                llvm::dbgs() << "found successor: \n";
+
+                auto succesorInputs = successor.getSuccessorInputs();
+                for (auto [idx, input] : llvm::enumerate(succesorInputs)) {
+                  llvm::dbgs() << "found successor input " << input
+                               << " with index " << idx << "\n";
+                  input.setType(op->getOperandTypes()[idx]);
```

Comment:
this I don't know: does RegionBranchOpInterface guarantee these types must agree?

Reply:
this should be op.getEntrySuccessorOperands(successor) instead of just op.getOperands/op.getOperandsTypes, then it is indeed guaranteed by the interface verifier.